File: backend/parse_linkedin.py
import re

_URN_PATTERNS = [
    re.compile(r"activity[:\-](\d{15,22})", re.IGNORECASE),
    re.compile(r"ugcPost[:\-](\d{15,22})", re.IGNORECASE),
    re.compile(r"share[:\-](\d{15,22})", re.IGNORECASE),
]


def extract_activity_id(url: str) -> str | None:
    if not url:
        return None
    for pattern in _URN_PATTERNS:
        match = pattern.search(url)
        if match:
            return match.group(1)
    return None


def is_valid_linkedin_post_url(url: str) -> bool:
    if not url:
        return False
    return "linkedin.com" in url.lower() and extract_activity_id(url) is not None


def build_embed_url(activity_id: str) -> str:
    return f"https://www.linkedin.com/embed/feed/update/urn:li:activity:{activity_id}"

File: frontend/src/utils/parseLinkedin.js
// Mirror of backend/parse_linkedin.py — keep the regexes in sync.
const URN_PATTERNS = [
  /activity[:\-](\d{15,22})/i,
  /ugcPost[:\-](\d{15,22})/i,
  /share[:\-](\d{15,22})/i,
];

export const extractActivityId = (url) => {
  if (!url) return null;
  for (const pattern of URN_PATTERNS) {
    const match = url.match(pattern);
    if (match) return match[1];
  }
  return null;
};

export const isValidLinkedInPostUrl = (url) => {
  if (!url) return false;
  return url.toLowerCase().includes('linkedin.com') && !!extractActivityId(url);
};

export const buildEmbedUrl = (activityId) =>
  `https://www.linkedin.com/embed/feed/update/urn:li:activity:${activityId}`;
